packages/api/portfolios/environments/updateEnvironment.ts

```diff
@@ -0,0 +1,78 @@
+import "reflect-metadata";
+import { createConnection } from "../../utils/database";
+import { Portfolio } from "../../../orm/entity/Portfolio";
+import { Application } from "../../../orm/entity/Application";
+import { Environment, IEnvironment } from "../../../orm/entity/Environment";
+import { EnvironmentRepository } from "../../repository/EnvironmentRepository";
+import { APIGatewayProxyResult, Context } from "aws-lambda";
+import { ApiSuccessResponse, SuccessStatusCode } from "../../utils/response";
+import internalSchema = require("../../models/internalSchema.json");
+import middy from "@middy/core";
+import xssSanitizer from "../../portfolioDrafts/xssSanitizer";
+import jsonBodyParser from "@middy/http-json-body-parser";
+import validator from "@middy/validator";
+import JSONErrorHandlerMiddleware from "middy-middleware-json-error-handler";
+import cors from "@middy/http-cors";
+import { ApiGatewayEventParsed } from "../../utils/eventHandlingTool";
+import { CORS_CONFIGURATION } from "../../utils/corsConfig";
+import { wrapSchema } from "../../utils/schemaWrapper";
+import { errorHandlingMiddleware } from "../../utils/errorHandlingMiddleware";
+import { IpCheckerMiddleware } from "../../utils/ipLogging";
+import { validateRequestShape } from "../../utils/shapeValidator";
+import { uniqueNameValidator } from "../../utils/businessRulesValidation";
+
+/**
+ * Submits an update to an environment of an application
+ *
+ * @param event - The PUT request from API Gateway
+ */
+export async function baseHandler(
+  event: ApiGatewayEventParsed<IEnvironment>,
+  context?: Context
+): Promise<APIGatewayProxyResult> {
+  const setupResult = validateRequestShape<IEnvironment>(event);
+  const { portfolioId, applicationId, environmentId } = setupResult.path;
+  const requestBody = setupResult.bodyObject;
+
+  // set up database connection
+  const connection = await createConnection();
+  let response: Environment;
+
+  try {
+    // ensures the portfolio, application, and environment id exists
+    await connection.getRepository(Portfolio).findOneOrFail({
+      id: portfolioId,
+    });
+    const application = await connection.getRepository(Application).findOneOrFail({
+      id: applicationId,
+    });
+    const environment = await connection
+      .getCustomRepository(EnvironmentRepository)
+      .findOneOrFail({ id: environmentId });
+
+    const environmentNames = await connection
+      .getCustomRepository(EnvironmentRepository)
+      .getAllEnvironmentNames(application.id);
+
+    // Throws error if duplicate name found
+    uniqueNameValidator(requestBody.name, environmentNames);
+
+    response = await connection
+      .getCustomRepository(EnvironmentRepository)
+      .updateEnvironment(environment.id, requestBody);
+    console.log("Updated Environment: " + JSON.stringify(response));
+  } finally {
+    connection.close();
+  }
```

Comment:
How do you stop the validator from throwing an error if you don't update the name in the request? Say fish environment exists in the db, and you want to update fish environment. how do you prevent validation from throwing an error if you don't rename fish environment (but update something else)?

Reply:
I believe this is captured by using `internalSchema.AppEnvAccess` in the validator. Only the `operators` should be updated according to the Spec. I believe the `PUT` operator handles changes to the environment name.